### PR TITLE
Address issues with unexpected MODULE values

### LINKED ISSUE #1591: Problems with MODULE
If the `MODULE` env variable is not set cocotb fails hard by raising (an `ImportError` of all things...) in the startup routine. It seems to me it should just run and exit with no failure because there are no tests to run.

`MODULE` might also accidentally contain trailing or leading separators, this isn't handled, and tests show it fails in an ugly way. `HOOKS`, whose usefulness is debatable, suffers the same problem.

### PR DESCRIPTION
**Pull request by ktbarrett**
Fixes #1591. Also addresses issue with `HOOKS`.

### CODE REVIEW

**Comment by ktbarrett** [comment]
Added support for trailing and leading whitespace in elements.

**Comment by imphil** [comment]
Code looks good modulo the empty MODULE error message change. Let's discuss that in https://github.com/cocotb/cocotb/issues/1591 and then return to this PR.

**Comment by eric-wieser** [comment]
I put some comments on  #1591.

**Comment by eric-wieser** [comment]
A concern orthogonal to the discussion there - I'm not convinced naming tests by their issue numbers are useful, unless the corner case is hard to describe concisely. More useful test names would be:

* `test_module_var_unset`
* `test_module_var_empty`
* `test_module_without_tests`

**Comment by ktbarrett** [comment]
Addressed comments here and in the linked issue.

If module is not set, we throw an error and exist uncleanly (without result.xml).
If module is empty (or no tests were found in any of the modules), we warn that no tests were run and exit cleanly (with result.xml).

Because of https://github.com/cocotb/cocotb/issues/1591#issuecomment-610414495, I removed the test for unset `MODULE` because the makefiles make testing that impossible right now.

**Comment by ktbarrett** [comment]
@imphil We can start returning non-0 exit codes once we move to something like #1562. The regression system will have control start to finish and the simulator will sit in the middle.

**Comment by imphil** [comment]
Yeah, that's a whole different topic.

File: cocotb/__init__.py [hunk 1]
```diff
@@ -164,12 +164,12 @@ def _initialise_testbench(root_name):
     test_str = os.getenv('TESTCASE')
     hooks_str = os.getenv('COCOTB_HOOKS', '')
 
-    if not module_str:
+    if module_str is None:
         raise ImportError("Environment variables defining the module(s) to " +
                           "execute not defined.  MODULE=\"%s\"" % (module_str))
```
Comment (eric-wieser) [inline cocotb/__init__.py:169]: This is dumb, if the module string was not defined then there's no poitn printing it, especially if we're going to imply it was set to the string `None`.
```suggestion
        raise ImportError("Environment variables MODULE which defines the module(s) to 
                          "execute is not defined.")
```

File: cocotb/regression.py [hunk 2]
```diff
@@ -180,6 +181,10 @@ def initialise(self):
             for thing in vars(module).values():
                 if hasattr(thing, "im_test"):
                     self._init_test(thing)
+                    have_tests = True
+
+        if not have_tests:
```
Comment (eric-wieser) [inline cocotb/regression.py:186]: Could you use `if self._queue` here? Or does that not work?

File: tests/test_cases/test_module_var_empty/Makefile [hunk 3]
```diff
@@ -0,0 +1,6 @@
+include ../../designs/sample_module/Makefile
+
+# test MODULE is empty
+
+MODULE := ""
+
```
Comment (eric-wieser) [inline tests/test_cases/test_module_var_empty/Makefile:6]: nit:
```suggestion
```

File: tests/test_cases/test_module_without_tests/Makefile [hunk 4]
```diff
@@ -0,0 +1,5 @@
+include ../../designs/sample_module/Makefile
+
+MODULE=test_nothing
+
+# test MODULE is set, but there are no tests across any of the modules
```
Comment (eric-wieser) [inline tests/test_cases/test_module_without_tests/Makefile:5]: nit, to match the others:
```suggestion
# test MODULE is set, but there are no tests across any of the modules

MODULE=test_nothing
```

File: cocotb/regression.py (hunk 2 above)
Comment (ktbarrett) [reply cocotb/regression.py:186]: Thing are only added to the queue if they initialize properly and are not skipped. What we want to see is if we find anything marked as a `test`.

File: cocotb/__init__.py (hunk 1 above)
Comment (ktbarrett) [reply cocotb/__init__.py:169]: I feel like `ImportError` is also improper, will change to `ValueError`.

File: cocotb/regression.py (hunk 2 above)
Comment (eric-wieser) [reply cocotb/regression.py:186]: I'd perhaps argue we want a warning for either case, but I'm happy to call that out of scope.

File: cocotb/__init__.py (hunk 1 above)
Comment (eric-wieser) [reply cocotb/__init__.py:169]: Hadn't even noticed that, good idea.

File: cocotb/__init__.py [hunk 5]
```diff
@@ -164,12 +164,11 @@ def _initialise_testbench(root_name):
     test_str = os.getenv('TESTCASE')
     hooks_str = os.getenv('COCOTB_HOOKS', '')
 
-    if not module_str:
-        raise ImportError("Environment variables defining the module(s) to " +
-                          "execute not defined.  MODULE=\"%s\"" % (module_str))
+    if module_str is None:
+        raise ValueError("Environment variable MODULE, which defines the module(s) to execute, is not defined.")
```
Comment (eric-wieser) [inline cocotb/__init__.py:168]: Line's a little long, but whatever

**Comment by ktbarrett** [comment]
So this is weird.

https://travis-ci.org/github/cocotb/cocotb/jobs/672124226#L4055-L4083

`MODULE` is set to the empty string, but the "MODULE not set" error is raised, indicating that `getenv` returned `None` instead of the empty string. Does `getenv` or the environment work differently on Windows? I don't know much of anything about Windows.

**Comment by eric-wieser** [comment]
Sounds like you ran into https://github.com/msys2/MINGW-packages/issues/1599

**Comment by eric-wieser** [comment]
CI failure is #1603

**Comment by eric-wieser** [comment]
Does this justify a newsfragment? In that whitespace is now stripped in the `MODULE` variable, meaning it can be wrapped across multiple lines?

**Comment by ktbarrett** [comment]
@eric-wieser I'll do some additional cleanup on other env var parsing in a later PR and I'll make a note then saying they are all whitespace and empty element tolerant.
